### ISSUE TITLE
 Ibm1130 (Windows GUI): Change position, color, text of buttons to match real 1130

### ISSUE THREAD

**Issue by Jim-F** [simh/simh#459]
- #### Context
There's a very nice on-line closeup image of a real 1130 console keyboard, courtesy of
Carl Claunch:

![1130_keyboard](https://cloud.githubusercontent.com/assets/17226010/26182866/fde904fc-3b48-11e7-98da-f25d9f984710.JPG)

This provides an opportunity to change the SimH 1130's GUI to more closely match the
real thing.

Current GUI:

![oldgui](https://cloud.githubusercontent.com/assets/17226010/26182885/2a1ab822-3b49-11e7-9278-5bfeacf172f5.png)

Modified GUI, more closely matching photo:

![newgui](https://cloud.githubusercontent.com/assets/17226010/26182893/3cf00e84-3b49-11e7-9d68-2f65a12a5a65.png)

These are simple changes to two locations in ibm1130_gui.c

First, renumber a list of macro definitions for the buttons/indicators so they'll still match
the positions in the graphic.

Old button macros:

```
--------------------------
#include <windows.h>

#define BUTTON_WIDTH  90
#define BUTTON_HEIGHT 50

#define IDC_KEYBOARD_SELECT     0
#define IDC_DISK_UNLOCK         1
#define IDC_RUN                 2
#define IDC_PARITY_CHECK        3
#define IDC_UNUSED              4
#define IDC_FILE_READY          5
#define IDC_FORMS_CHECK         6
#define IDC_POWER_ON            7
#define IDC_POWER               8
#define IDC_PROGRAM_START       9
#define IDC_PROGRAM_STOP        10
#define IDC_LOAD_IAR            11
#define IDC_KEYBOARD            12
#define IDC_IMM_STOP            13
#define IDC_RESET               14
#define IDC_PROGRAM_LOAD        15
--------------------------
```

New button macros:

```
--------------------------
#include <windows.h>

#define BUTTON_WIDTH  90
#define BUTTON_HEIGHT 50

#define IDC_UNUSED              0            <--- changed
#define IDC_DISK_UNLOCK         1
#define IDC_RUN                 2
#define IDC_KEYBOARD_SELECT     3            <---
#define IDC_POWER_ON            4            <---
#define IDC_FILE_READY          5
#define IDC_PARITY_CHECK        6            <---
#define IDC_FORMS_CHECK         7            <---
#define IDC_POWER               8
#define IDC_PROGRAM_START       9
#define IDC_PROGRAM_STOP        10
#define IDC_LOAD_IAR            11
#define IDC_KEYBOARD            12
#define IDC_IMM_STOP            13
#define IDC_RESET               14
#define IDC_PROGRAM_LOAD        15
--------------------------
```

Then change their properties (position, color, and/or label).  If a button has changed
location, be sure to carry the initial "state" boolean value along with it.

Old button properties:

```
--------------------------
} btn[] = {
    0, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "KEYBOARD\nSELECT",     FALSE,  FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    0, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "DISK\nUNLOCK",         FALSE,  TRUE,   RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    0, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "RUN",                  FALSE,  FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    0, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "PARITY\nCHECK",        FALSE,  FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,

    1, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "",                     FALSE,  FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    1, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "FILE\nREADY",          FALSE,  FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    1, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "FORMS\nCHECK",         FALSE,  FALSE,  RGB(255,255,0),     NULL, NULL, NULL,   TRUE,
    1, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "POWER\nON",            FALSE,  TRUE,   RGB(255,255,180),   NULL, NULL, NULL,   TRUE,

    2, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "POWER",                TRUE,   FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    2, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nSTART",       TRUE,   FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    2, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nSTOP",        TRUE,   FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,
    2, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "LOAD\nIAR",            TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,

    3, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "KEYBOARD",             TRUE,   FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    3, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "IMM\nSTOP",            TRUE,   FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,
    3, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "CHECK\nRESET",         TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,
    3, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nLOAD",        TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,
--------------------------
```

New button properties:

```
--------------------------
} btn[] = {
    0, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "",                     FALSE,  FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,  <--- changed
    0, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "DISK\nUNLOCK",         FALSE,  TRUE,   RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    0, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "RUN",                  FALSE,  FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    0, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "K B\nSELECT",          FALSE,  FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,  <---

    1, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "POWER\nON",            FALSE,  TRUE,   RGB(255,255,180),   NULL, NULL, NULL,   TRUE,  <---
    1, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "FILE\nREADY",          FALSE,  FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    1, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "PARITY\nCHECK",        FALSE,  FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,  <---
    1, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "FORMS\nCHECK",         FALSE,  FALSE,  RGB(255,255,0),     NULL, NULL, NULL,   TRUE,  <---

    2, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "POWER",                TRUE,   FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    2, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nSTART",       TRUE,   FALSE,  RGB(0,255,0),       NULL, NULL, NULL,   TRUE,
    2, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nSTOP",        TRUE,   FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,
    2, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "LOAD\nIAR",            TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,

    3, 0, BUTTON_WIDTH, BUTTON_HEIGHT,  "KEYBOARD",             TRUE,   FALSE,  RGB(255,255,180),   NULL, NULL, NULL,   TRUE,
    3, 1, BUTTON_WIDTH, BUTTON_HEIGHT,  "IMM\nSTOP",            TRUE,   FALSE,  RGB(255,0,0),       NULL, NULL, NULL,   TRUE,
    3, 2, BUTTON_WIDTH, BUTTON_HEIGHT,  "RESET",                TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,  <--- label changed
    3, 3, BUTTON_WIDTH, BUTTON_HEIGHT,  "PROGRAM\nLOAD",        TRUE,   FALSE,  RGB(0,0,255),       NULL, NULL, NULL,   TRUE,
--------------------------
```

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue
Same as before.
- #### how you built the simulator or that you're using prebuilt binaries

Same as before
- #### the simulator configuration file (or commands) which were used when the problem occurred.

Same as before.
- #### the expected behavior and the actual behavior

- #### you may also need to provide specific pointers to data files that may be necessary to demonstrate the problem

**Comment by markpizz** [comment]
This looks almost fine.  

What else would it take to change the text color on the Red Green and Blue buttons to be White?

**Comment by markpizz** [comment]
I suggested white due to the picture above, but the only one which really needs to be changed is the black on the Blue buttons.

**Comment by Jim-F** [comment]
> This looks almost fine.
>
> What else would it take to change the text color on the Red Green and Blue buttons to be White?

Thanks for indulging me on this!  There might indeed be other things that could be done
to make the buttons and indicators even more authentic looking -- adjusting the colors
would be fairly easy, changing the text color might be possible (I haven't looked into
it in detail), and of course there's the possibility of adding those black or white
stripes at the top and bottom of the indicators on the left.

I'll leave this open, then, and experiment with it some more.  (Authenticity is important,
even if it's just cosmetic!)

**Comment by Jim-F** [comment]
You know, it crossed my mind that the way to do this might be to ask Mr. Claunch
to provide high-resolution closeup photos of the buttons and the indicators
(the latter in both illuminated and unilluminated states).  These could then be
changed to .bmp images of the right size to be loaded into subwindows of the
main window as needed.

The "aspect ratios" of the real buttons/indicators aren't quite the same as the
shapes of those on the existing GUI, but leaving a little more black around more
realistically-shaped photo-based buttons/indicators, instead of having them
abut as they do now, might also add to a more authentic look.  Starting with a photograph
as the basis for a GUI isn't new -- that's what Jörg Hoppe did for his DEC
front panels for SimH.

In fact, if we **really** wanted to get ambitious, we could add a bit of animation
and sound for when a button is pushed -- have it look like it's actually being
depressed when the mouse button is depressed and then springing back
when the mouse button is released, and making a soft click each time.

And we could even add a "lamp test" command (or button, somewhere),
just to be able to see indicators illuminated that would otherwise never
be seen lit up (PARITY CHECK, FORMS CHECK).

**Comment by Jim-F** [comment]
> And we could even add a "lamp test" command (or button, somewhere). . .

In fact, while I'm at it, I could add all those cycle-level indicators that aren't seen
because SimH can't operate them, but they could show up just for the 'lamp test".
And we could make the register indicators look a little more realistic, (again,
maybe based on photographs).  The whole shape of the GUI would have to
change -- it would have to be wider, more like a real 1130 panel.  Maybe we could
have alternative builds so that anybody who wanted to keep the GUI as is
could do so.

This could be quite a project!

**Comment by markpizz** [comment]
Keep changes to ibm1130_gui.c and DO NOT add additional library dependencies.

As for going wider, be careful to not go too wide that you no longer have room for an explorer window that you can 'drag and drop' files from.

I wouldn't worry about keeping the legacy look.

**Comment by Jim-F** [comment]
Mark wrote (in a comment on issue #463):

> All of the recent work on the Windows specific GUI
> for the ibm1130 is effort spent which can not be leveraged
> anywhere else.
>
> Additionally there exists an interface for simulators to
> provide front panels (GUI/software ones and real hardware ones)
> using the sim_frontpanel API. The frontpanel APIs allow someone
> who wants to develop a front panel to do that without messing
> with the internals of a simulator.

In that case, going beyond the simple changes we've already made
to the buttons and indicators in the more-or-less schematic
existing 1130 GUI is probably not a wise pursuit.

Anything fancier than what exists now is probably best left
to a front panel a la Joerg Hoppe, or indeed left to a
specialized cycle-accurate simulator rather than Simh
(which I gather is once again being tackled for the
IBM 1620 at the Computer History Museum, according to a recent
post on CCTalk:
http://www.classiccmp.org/pipermail/cctalk/2017-April/033937.html ).

So I'll go ahead and close this issue.

**Comment by markpizz** [comment]
Jim,

If you do the work here, and, as I suggested, keep the changes to ibm1130_gui.c without adding additional external library dependencies, I'll be glad to take it.  This will be minimal impact to me while adding polish to the existing solution.  If you're willing, reopen this issue and press on.

- Mark

**Comment by Jim-F** [comment]
Mark,

You wrote:

> If you do the work here, and, as I suggested, keep the changes to ibm1130_gui.c without adding additional external library dependencies, I'll be glad to take it. This will be minimal impact to me while adding polish to the existing solution. If you're willing, reopen this issue and press on.

Well, I suspect that to do what I was thinking of doing (loading
and unloading images and so forth), I'd want to use libgd -- which
ibm1130 already depends on, but only for the plotter (which isn't
enabled by default -- but then again, a fancy[er] GUI needn't be
enabled by default either).

I have no experience with graphics libraries -- SDL might be able
to do the same thing, but then we'd have ibm1130 depending on both
SDL (which is allowed, I gather) **and** libgd (which isn't really
allowed, and is only there because of the legacy plotter stuff).

I don't want to ruffle any feathers here.

It's not like I've already started work on this, and I suspect
you might be right that a really fancy GUI belongs in something
**outside** of the regular SimH package -- like Joerg Hoppe's
panels (which are really spectacular).

In fact, the 1130 really begs for cycle-level emulation, to blink
**all** the lights, like Dave Babcock is now apparently revisiting
for the 1620.  It's a shame that the work Richard Jowsey started
on a cycle-accurate 1620 GUI emulator in Java for the CHM 1620 restoration
back in 2001 was never completed.  I still run that occasionally
(running the CU-01 diagnostic program) just to watch the lights
blink.  In fact, I showed it off at a party (on a big TV screen)
just this past weekend, along with the 1130, a PDP-11 with
Hoppe's front panel, and Bernhard Baehr's PDP-8 running in a
hackintosh VMware image.  The old computers were a big hit, I think.

Jim F.

**Comment by markpizz** [comment]
I was thinking that if you stuck with .bmp's which are already part of what is going on with native Windows API's you could get there pretty easily.  It's not like you're trying to get active animation, right?  Just changing the image that a particular button is displayed with that doesn't change with any significant frequency.

**Comment by Jim-F** [comment]
Mark,

You wrote:

> I was thinking that if you stuck with .bmp's which are already
> part of what is going on with native Windows API's you could get
> there pretty easily. It's not like you're trying to get active
> animation, right? Just changing the image that a particular button
> is displayed with that doesn't change with any significant frequency.

I'll experiment a little, and if I come up with anything that
seems worthwhile, I'll open a new issue for "photo-realistic buttons/indicators",
or whatever, and submit what I think I've accomplished.

If I can get smoothly operating code that doesn't break the rules
(using cat pictures or whatever sample images I have on hand ;-> ),
**then** I can think about bothering Carl Claunch to take photos of
the buttons and lights on his real machine.

I was also a little worried that there may be a kind of "uncanny valley"
where you've added a few somewhat more realistic visual elements to what is
basically a pretty schematic display, and created a gestalt that's actually
uglier and/or more disconcerting.  But maybe it's just a matter of whatever
you've gotten used to.  Maybe cat picture would work, too.  ;->

Jim F.
